### PR TITLE
implement complex div/mul libcalls

### PR DESCRIPTION
**Pull request by folkertdev**
This needs https://github.com/rust-lang/rust/pull/162832 to be merged and synced for the `Div`/`Mul` instances.

Maybe we can iterate on the implementation already, or if not this will just sit for a couple of days.

### CODE REVIEW

File: src/shims/complex_numbers.rs
Comment (RalfJung) [inline src/shims/complex_numbers.rs:1]: This feels like it should just go into shims/math.rs

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: src/shims/complex_numbers.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    }
+}
+
+type ComplexFloatBinop<S> =
+    fn(a: IeeeFloat<S>, b: IeeeFloat<S>, c: IeeeFloat<S>, d: IeeeFloat<S>) -> Complex<IeeeFloat<S>>;
+
+fn complex_binop<'tcx, S: Semantics>(
+    this: &mut MiriInterpCx<'tcx>,
+    op: ComplexFloatBinop<S>,
+    [a, b, c, d]: &[OpTy<'tcx>; 4],
```
Comment (RalfJung) [inline src/shims/complex_numbers.rs]: These names are not very helpful... at the very least please document them. But I assume they correspond to the real and imaginary parts of the left and right input, respectively -- the names should reflect that to some extend, like `[a_re, a_im, b_re, b_im]` or so.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: src/shims/complex_numbers.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    mut b: IeeeFloat<S>,
+    mut c: IeeeFloat<S>,
+    mut d: IeeeFloat<S>,
+) -> Complex<IeeeFloat<S>> {
+    let ac = (a * c).value;
+    let bd = (b * d).value;
+    let ad = (a * d).value;
+    let bc = (b * c).value;
+
+    let z = Complex::new((ac - bd).value, (ad + bc).value);
```
Comment (RalfJung) [inline src/shims/complex_numbers.rs]: Isn't this duplicating naive_complex_mul?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: src/shims/complex_numbers.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    let zero_if_nan =
+        |x: IeeeFloat<S>| if x.is_nan() { IeeeFloat::<S>::ZERO.copy_sign(x) } else { x };
+
+    let signed_unit_if_inf = |x: IeeeFloat<S>| {
+        let mag =
+            if x.is_infinite() { IeeeFloat::<S>::from_u128(1).value } else { IeeeFloat::<S>::ZERO };
+        mag.copy_sign(x)
+    };
+
+    let mut recalc = false;
```
Comment (RalfJung) [inline src/shims/complex_numbers.rs]: Why is this so imperative? What's the point of this mutable variable?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: src/shims/complex_numbers.rs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    let bc = (b * c).value;
+
+    Complex::new((ac - bd).value, (ad + bc).value)
+}
+
+/// Returns the quotient of `(a + ib)` and `(c + id)`.
+///
+/// This implementation uses the standard formula, but has special behavior when the output
+/// of that formula has both a real and imaginary component that are NaN.
+fn complex_div<S: Semantics>(
```
Comment (RalfJung) [inline src/shims/complex_numbers.rs]: What's the standard formula? What's the special behavior?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: src/helpers.rs [hunk 5]
```diff
@@ -170,6 +170,19 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         try_resolve_did(*self.eval_context_ref().tcx, path, None).is_some()
     }
 
+    fn complex_ty_layout(&self, component_type: Ty<'tcx>) -> TyAndLayout<'tcx> {
```
Comment (RalfJung) [inline src/helpers.rs:173]: Please add a doc comment

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: tests/pass/shims/complex_numbers.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
+#![feature(complex_numbers)]
```
Comment (RalfJung) [inline tests/pass/shims/complex_numbers.rs:1]:
```suggestion
//@ run-native
#![feature(complex_numbers)]
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: tests/pass/shims/complex_numbers.rs
Comment (RalfJung) [inline tests/pass/shims/complex_numbers.rs:1]: Usually we like to have tests that directly invoke the functions we shim, because libcore may change and existing tests may suddenly stop covering our shims.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: src/shims/complex_numbers.rs [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+    if !recalc {
+        return z;
+    }
+
+    let mut z = naive_complex_mul(a, b, c, d);
+    z.re = (z.re * IeeeFloat::<S>::INFINITY).value;
+    z.im = (z.im * IeeeFloat::<S>::INFINITY).value;
+
+    z
```
Comment (RalfJung) [inline src/shims/complex_numbers.rs]: There's a lot going on in this function and none of it is explained.  What exactly is the spec being implemented here? Why is "at least one non-NaN" sufficient to know that the "fast-path" is fine? What is the point of all the logic after that NaN check?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/a326541ee8ac82141546a15aea2e5abf1e09c19a..d48a83132a18e90a44145f24c052343378c56714)*

File: src/shims/complex_numbers.rs (hunk 2 above)
Comment (folkertdev) [reply src/shims/complex_numbers.rs]: yes, but the `ac` etc. variables are used below.

File: src/shims/complex_numbers.rs (hunk 3 above)
Comment (folkertdev) [reply src/shims/complex_numbers.rs]: It is translated directly from compiler-rt, e.g. 

https://github.com/llvm/llvm-project/blob/5bff640ac99dc29ab61cb55bdfc9b992d09b5a85/compiler-rt/lib/builtins/muldc3.c

File: src/shims/complex_numbers.rs (hunk 7 above)
Comment (folkertdev) [reply src/shims/complex_numbers.rs]: This is specified in annex G of the C23 standard 

https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf#page=556

I'll port over the comments, those are nice.

Comment (RalfJung) [reply src/shims/complex_numbers.rs]: So I guess this is about this part of the spec?

image

Please explicitly explain how the code achieves that spec. Tests should also reference that spec to ensure they cover all these corner cases.

Comment (RalfJung) [reply src/shims/complex_numbers.rs]: That document also says

>  It is less than ideal in that undue overflow and underflow could occur.

What is that about?

File: src/shims/math.rs [hunk 8]
```diff
@@ -249,9 +252,254 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+}
+
+/// Returns the product of `a + ib` and `c + id`.
+///
+/// The standard formula is `(ac - bd) + (ad + bc)i`, but this function has custom behavior when
+/// This implementation uses the standard formula:
+///
+/// a+bi * c+di = ((ac + bd) + (bc - ad)i) / (c*c + d*d)
+///
+/// But with recovery of infinities if the above expresion results in NaN + NaNi.
```
Comment (RalfJung) [inline src/shims/math.rs:353]: Please explain what "recovery of infinities" means, or reference a document that explains this.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/99cc4400b5ccc58848d75261741e6e9e998468de..2a59386125e572f1c0e6b3ffe2f45e4a33bf8dbf)*

File: src/shims/math.rs [hunk 9]
```diff
@@ -249,9 +252,254 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+fn complex_mul<S: Semantics>(
+    mut a: IeeeFloat<S>,
+    mut b: IeeeFloat<S>,
+    mut c: IeeeFloat<S>,
+    mut d: IeeeFloat<S>,
+) -> Complex<IeeeFloat<S>> {
+    let ac = (a * c).value;
+    let bd = (b * d).value;
+    let ad = (a * d).value;
+    let bc = (b * c).value;
```
Comment (RalfJung) [inline src/shims/math.rs:363]: These names are still completely meaningless.^^

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/99cc4400b5ccc58848d75261741e6e9e998468de..2a59386125e572f1c0e6b3ffe2f45e4a33bf8dbf)*

File: src/shims/math.rs [hunk 10]
```diff
@@ -249,9 +252,254 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+    let res = op(a, b, c, d);
+
+    this.write_scalar(res.re, &this.project_field(dest, FieldIdx::ZERO)?)?;
+    this.write_scalar(res.im, &this.project_field(dest, FieldIdx::ONE)?)?;
+    interp_ok(())
+}
+
+/// Returns the product of `a + ib` and `c + id`.
+///
+/// The standard formula is `(ac - bd) + (ad + bc)i`, but this function has custom behavior when
```
Comment (RalfJung) [inline src/shims/math.rs:348]: Sentence seems to cut off at the end here.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/99cc4400b5ccc58848d75261741e6e9e998468de..2a59386125e572f1c0e6b3ffe2f45e4a33bf8dbf)*

File: src/shims/math.rs [hunk 11]
```diff
@@ -249,9 +252,254 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+        recalc = true;
+    }
+
+    if !recalc {
+        return z;
+    }
+
+    let mut z = naive_complex_mul(a, b, c, d);
+    z.re = (z.re * IeeeFloat::<S>::INFINITY).value;
+    z.im = (z.im * IeeeFloat::<S>::INFINITY).value;
```
Comment (RalfJung) [inline src/shims/math.rs:419]: What is this doing?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/99cc4400b5ccc58848d75261741e6e9e998468de..2a59386125e572f1c0e6b3ffe2f45e4a33bf8dbf)*

File: src/shims/math.rs [hunk 12]
```diff
@@ -249,9 +252,254 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+        |x: IeeeFloat<S>| if x.is_nan() { IeeeFloat::<S>::ZERO.copy_sign(x) } else { x };
+
+    let signed_unit_if_inf = |x: IeeeFloat<S>| {
+        let mag =
+            if x.is_infinite() { IeeeFloat::<S>::from_u128(1).value } else { IeeeFloat::<S>::ZERO };
+        mag.copy_sign(x)
+    };
+
+    // Recover infinities that computed as NaN + iNaN.
+    let mut recalc = false;
```
Comment (RalfJung) [inline src/shims/math.rs:382]: This comment looks like it should be explaining that variable.
```suggestion
    // Recover infinities that computed as NaN + iNaN.

    // Will be set to `true` if the double-NaN result is wrong and
    // we have to do infinity recovery.
    let mut recalc = false;
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/99cc4400b5ccc58848d75261741e6e9e998468de..2a59386125e572f1c0e6b3ffe2f45e4a33bf8dbf)*

File: src/shims/math.rs [hunk 13]
```diff
@@ -249,9 +252,254 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+    mut d: IeeeFloat<S>,
+) -> Complex<IeeeFloat<S>> {
+    // Scale the inputs
+    let max = IeeeFloat::<S>::max(c.abs(), d.abs());
+    let mut neg_ilogbw = 0;
+    if max.is_finite() && max != IeeeFloat::<S>::ZERO {
+        neg_ilogbw = max.ilogb().checked_neg().unwrap();
+        c = c.scalbn(neg_ilogbw);
+        d = d.scalbn(neg_ilogbw);
+    }
```
Comment (RalfJung) [inline src/shims/math.rs:461]: What is happening here? Scale the inputs how and why?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/99cc4400b5ccc58848d75261741e6e9e998468de..2a59386125e572f1c0e6b3ffe2f45e4a33bf8dbf)*

Comment (folkertdev) [reply src/shims/math.rs]: My theory is that it's about preventing overflow. I'll have to look into that more though.

File: src/shims/math.rs (hunk 9 above)
Comment (folkertdev) [reply src/shims/math.rs:372]: They represent `a + bi` and `c + di`, and `ac` is the product `a * c`? They are abstract but relate to the doc comment and this is, I think, the standard way to write this down?

Comment (folkertdev) [reply src/shims/math.rs:372]: in other words: I don't think using `lhs_re`, `lhs_re_mul_rhs_re` etc is substantially clearer.

Comment (RalfJung) [reply src/shims/math.rs:372]: Hm, I guess we're going to link to https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf#page=556 which also uses a, b, c, d...

File: src/shims/math.rs [hunk 14]
```diff
@@ -249,9 +252,255 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+    };
+
+    // Recover infinities that computed as NaN + iNaN.
+
+    // Will be set to `true` if the double-NaN result is wrong and we have to do infinity recovery.
+    let mut recalc = false;
+
+    if a.is_infinite() || b.is_infinite() {
+        // Replace infinities with a signed unit value, and NaN with zero.
+        // This avoids invalid infinity arithmetic.
```
Comment (RalfJung) [inline src/shims/math.rs:397]: We're doing the inf replacement on a, b and the NaN replacement on c, d. Why that? The comment makes it sounds like we're doing both replacements on all values.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5319/70196c33af6b4221d0f310c4bdc5b6f6c46b3978..779d463a35d845f5b09f2183a6b1e6253b0b7ed2)*

File: src/shims/complex_numbers.rs (hunk 7 above)
Comment (folkertdev) [reply src/shims/complex_numbers.rs]: > It is less than ideal in that undue overflow and underflow could occur.

I found https://arxiv.org/pdf/1210.4539 which presents some cases where the C specification algorithm still produces incorrect results. There is also https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1399.htm which presents more incorrect results. It proposes an overhaul to complex division that has clearly not been adopted.

File: src/shims/math.rs (hunk 13 above)
Comment (folkertdev) [reply src/shims/math.rs]: Hmm, the most I could find about it is from 

> In the ISO/IEC 9899:TC3 C99 standard [9], section G.5.1, “Multiplicative operators”, the authors present a _Cdivd function in the C language which implements complex division. According to Kahan [10] (in the Appendix, “Over/Underflow Undermines Complex Number Division in Java”), this code is due to Jim Thomas and Fred Tydeman.
> 
> The algorithm is based on a prescaling of c and d, such that their base-2 exponent is made zero. This scales the denominator c² + d² and avoids most overflows or underflows. The post-scaling produces then the correct result. This scaling is based on a power of 2, which avoids rounding.
> 
> Only in the case of an IEEE exception, the algorithm recomputes the division, taking into account for NaNs and Infinities. The code does not defend against overflow and underflow in the calculation of the numerator.
> 
> We will show later in this paper that the C99 algorithm performs nearly as well as Smith’s algorithm, but fails in many cases.

File: src/shims/math.rs (hunk 8 above)
Comment (folkertdev) [reply src/shims/math.rs:352]: I've added some examples

File: src/shims/math.rs (hunk 9 above)
Comment (folkertdev) [reply src/shims/math.rs:372]: Yeah, I've linked that now. Having done more reading, it's really not "the best" way of implementing these operations (especially division), but for these libcalls the C reference algorithm is what we should implement.

File: src/shims/complex_numbers.rs (hunk 7 above)
Comment (RalfJung) [reply src/shims/complex_numbers.rs]: So far I don't even know what "correct" and "incorrect" mean. What is the *specification* for these operations? That's what the doc comment should explain. Without that, discussing correctness is pointless.

However it seems like correctness is not actually precisely defined and the functions are approximations. Doesn't that mean we should randomize them like we randomize sin, log, etc?
